File: adversary/src/Kusanagi/Lying.hs
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at https://mozilla.org/MPL/2.0/.
-- Copyright (c) 2026 2youg1 and the kusanagi contributors

{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Two lies a host can tell without forging anything.
--
-- "Kusanagi.Model" already flips a byte, which is the lie a host tells by being
-- broken. These are the two it can tell while every byte it holds is a byte an
-- endpoint really wrote and really signed, so no signature check can see them:
--
-- * **Transplant.** Serve the object from one address at another. The bytes are
--   genuine and the signature verifies; only the position is a lie. This network
--   answers it with the key rather than with a check, because an address derives
--   the key its contents are sealed under — so the question this property really
--   asks is whether that derivation is load-bearing, and it would catch the day
--   somebody makes the key depend on the segment instead.
--
-- * **Vanish.** Stop serving an object. Nothing can prevent this; a store that
--   will not hand bytes over is a store that will not hand bytes over. What must
--   not follow is a reader believing *less* than it has already verified,
--   because "she never sent the cancellation" is a lie a disappearance can tell.
--
-- Both are stated as relations. Neither says what the program should print; each
-- says how two runs must stand to one another.
module Kusanagi.Lying
  ( Written (..)
  , writeSome
  , transplantIsRefused
  , historyNeverShrinks
  ) where

import Data.Text qualified as Text
import Data.Word (Word64)

import Kusanagi.Answer
  ( Address
  , Answer (..)
  , ChannelName (..)
  , Outcome (..)
  )
import Kusanagi.Door (Door)
import Kusanagi.Door qualified as Door
import Kusanagi.Ground (Ground, transplant, vanish)

-- | A stocked channel, and where the host put every segment of it.
data Written = Written
  { writtenReader :: FilePath
  , writtenChannel :: ChannelName
  , writtenAddresses :: [Address]
  }
  deriving stock (Eq, Show)

-- | Opens a channel and writes @count@ segments, keeping every address.
--
-- The addresses come from the sender's own @--json@, not from listing the host's
-- directory. That matters: a property that learned addresses by looking at the
-- host would silently stop testing anything the day the host's layout changed.
writeSome :: Door -> FilePath -> FilePath -> FilePath -> Int -> IO Written
writeSome door writer reader host count = do
  minted <- Door.ask door writer (Door.Invite channel host Door.Forever Door.both)
  invitation <- case minted of
    Accepted (Invited _ line _) -> pure line
    other -> fail ("the invitation was refused: " <> show other)
  joined <- Door.ask door reader (Door.Join invitation channel)
  case joined of
    Accepted Joined {} -> pure ()
    other -> fail ("the channel could not be joined: " <> show other)
  addresses <- mapM say [1 .. count]
  pure
    Written
      { writtenReader = reader
      , writtenChannel = channel
      , writtenAddresses = addresses
      }
  where
    channel = ChannelName "peer"
    say n = do
      said <- Door.ask door writer (Door.Send channel (Text.pack ("segment " <> show n)))
      case said of
        Accepted (Sent _ _ address) -> pure address
        other -> fail ("a segment was refused: " <> show other)

-- | Bytes served at an address other than their own are not read as a segment.
--
-- The two addresses are both real drops of the same stream, so nothing about
-- the object is unusual: same author, same key length, same shape, written
-- minutes apart by the same endpoint. Only the height is wrong.
--
-- Success is either a refusal or a read that stops below the transplant. Both
-- are honest; what fails is a read that hands the caller a segment sitting at a
-- height its author never put it at.
transplantIsRefused :: Door -> Ground -> Written -> IO (Either String ())
transplantIsRefused door ground written =
  case writtenAddresses written of
    (first : second : _) -> do
      before <- heightOf door written
      case before of
        Left reason -> pure (Left ("the stream did not read back first: " <> reason))
        Right seen -> do
          transplant ground second first
          after <- readingOf door written
          pure $ case after of
            Refused _ -> Right ()
            Accepted (Read _ _ height _)
              | height < seen -> Right ()
              | otherwise ->
                  Left
                    ( "a segment moved to another height was read as though it \
                      \belonged there: height "
                        <> show height
                        <> " after the move, "
                        <> show seen
                        <> " before it"
                    )
            other -> Left ("a read answered with something else: " <> show other)
    _ -> pure (Right ())

-- | A reader that has verified a height is never talked down from it.
--
-- Read once, let the host drop everything above the floor, read again. The
-- second answer may fail and it may be short, but it must not report a lower
-- height than the reader had already checked for itself — that number is what an
-- agent polls from, and a host that can lower it can replay a conversation.
historyNeverShrinks :: Door -> Ground -> Written -> IO (Either String ())
historyNeverShrinks door ground written =
  case reverse (writtenAddresses written) of
    [] -> pure (Right ())
    (top : _) -> do
      before <- heightOf door written
      case before of
        Left reason -> pure (Left ("the stream did not read back first: " <> reason))
        Right seen -> do
          vanish ground top
          after <- readingOf door written
          pure $ case after of
            Refused _ -> Right ()
            Accepted (Read _ _ height _)
              | height >= seen -> Right ()
              | otherwise ->
                  Left
                    ( "the host deleted one object and walked a reader back from \
                      \height "
                        <> show seen
                        <> " to "
                        <> show height
                        <> "; a reader that has verified a height must not \
                           \believe less than it checked"
                    )
            other -> Left ("a read answered with something else: " <> show other)

-- | The height a read reports, or why it did not report one.
heightOf :: Door -> Written -> IO (Either String (Maybe Word64))
heightOf door written = do
  answer <- readingOf door written
  pure $ case answer of
    Accepted (Read _ _ height _) -> Right height
    other -> Left (show other)

readingOf :: Door -> Written -> IO Answer
readingOf door written =
  Door.ask door (writtenReader written) (Door.Read (writtenChannel written))

File: crates/kusanagi/tests/lying.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
// Copyright (c) 2026 2youg1 and the kusanagi contributors

//! Lies a host can tell while every byte it serves is genuine.
//!
//! `unlinkable.rs` asks what a host learns. This asks what a host can make
//! somebody believe. The difference matters because the lies here need no
//! forgery: the host holds real segments, really signed, and tells its lie by
//! choosing which ones to hand over and where.
//!
//! Found by `adversary/src/Kusanagi/Lying.hs`, which drives the shipped binary
//! and asserts that two readings stand in the right relation. Haskell searched
//! for it; this file is where the repository remembers it.

#![allow(
    clippy::unwrap_used,
    clippy::expect_used,
    clippy::panic,
    clippy::indexing_slicing,
    reason = "test code"
)]

mod common;

use common::{Endpoint, invite_line, json, scratch, stored};
use kusanagi::{Request, Whose};

/// Deletes whatever the host holds at `address`.
fn vanish(host: &std::path::Path, address: &str) {
    std::fs::remove_file(common::object_path(host, address)).expect("the host held nothing there");
}

#[test]
fn a_host_cannot_talk_a_reader_down_from_a_height_it_verified() {
    let ground = scratch("lying-rollback");
    let host = ground.join("host");
    let alice = Endpoint::new(ground.join("alice"));
    let bob = Endpoint::new(ground.join("bob"));

    let invitation = invite_line(&alice, "bob", &host.display().to_string());
    bob.run(&Request::Join {
        invite: invitation,
        name: "alice".to_owned(),
        habit: kusanagi::Habit::default(),
    })
    .unwrap();
    alice.send("bob", "the offer stands");
    let withdrawn = alice.send_reporting("bob", "on second thoughts, it does not");

    let read = |after| {
        bob.run(&Request::Read {
            name: "alice".to_owned(),
            after,
            whose: Whose::Peer,
        })
    };

    // Bob reads both, and now knows something the host would like him to forget.
    assert_eq!(json(&read(None).unwrap())["height"], 1);

    // The host drops the retraction. Nothing is forged: what remains is one real
    // segment, correctly signed, forming a perfect chain of length one. A reader
    // starting from nothing would accept it without a murmur, which is exactly
    // why an endpoint that has read before must not.
    vanish(&host, &withdrawn);
    let refused = read(None).unwrap_err();
    assert_eq!(refused.code(), "kusanagi.history_changed");
    assert!(
        refused.render(false).contains("doctor"),
        "the way out does not name the command that measures the host: {}",
        refused.render(false)
    );

    // A floor below the deleted segment still has to fetch it, so it fails the
    // same way. That is not a weakness of the floor; it is the caller asking for
    // a segment the host is refusing to serve.
    assert_eq!(
        read(Some(0)).unwrap_err().code(),
        "kusanagi.history_changed"
    );

    // The poll an agent runs — a floor at the height it already holds — refuses
    // by construction rather than by comparison: it never looks below what it
    // verified, so there is nothing down there for the host to take away.
    let polled = json(&read(Some(1)).unwrap());
    assert_eq!(
        polled["height"], 1,
        "a poll lost the height bob had verified"
    );
    assert!(polled["segments"].as_array().unwrap().is_empty());

    std::fs::remove_dir_all(&ground).ok();
}

#[test]
fn genuine_bytes_served_at_the_wrong_address_are_not_a_segment() {
    let ground = scratch("lying-transplant");
    let host = ground.join("host");
    let alice = Endpoint::new(ground.join("alice"));
    let bob = Endpoint::new(ground.join("bob"));

    let invitation = invite_line(&alice, "bob", &host.display().to_string());
    bob.run(&Request::Join {
        invite: invitation,
        name: "alice".to_owned(),
        habit: kusanagi::Habit::default(),
    })
    .unwrap();
    let first = alice.send_reporting("bob", "one");
    let second = alice.send_reporting("bob", "two");

    // Serve the second segment from the first segment's address. Both objects
    // are this endpoint's own, minutes apart, same author, same shape. Only the
    // height is a lie, and no signature can see it — the address is what carries
    // the answer, because the key the bytes are sealed under is derived from it.
    let held: Vec<(String, Vec<u8>)> = stored(&host);
    let body = held
        .iter()
        .find(|(address, _)| *address == second)
        .map(|(_, bytes)| bytes.clone())
        .expect("the host is not holding the second segment");
    std::fs::write(common::object_path(&host, &first), body).unwrap();

    let refused = bob
        .run(&Request::Read {
            name: "alice".to_owned(),
            after: None,
            whose: Whose::Peer,
        })
        .unwrap_err();
    // `seal.rejected`, not a chain or signature error: the bytes never became a
    // segment at all. Moving an object is answered one layer below verification,
    // by the key the address derives, which is why no check had to be written for
    // it and why removing that derivation would silently open this door.
    assert_eq!(
        refused.code(),
        "seal.rejected",
        "bytes moved to another height were not stopped by the key: {}",
        refused.render(false)
    );

    std::fs::remove_dir_all(&ground).ok();
}
